distsqlrun: turn "flow already registered" into an error

andreimatei opened the pull request:
Two flows with the same id seem to be scheduled on the same node which
used to cause panics. How that can happen is currently unclear; cosmic
rays? This patch turns the panic into an RPC or query error (depending
on sync/async flow), adds some more sanity checks and adds a
representation of the flow to the error and the error will also invite
users to report on #12876.

Touches #12876.

andreimatei committed:
distsqlrun: turn "flow already registered" into an error

Two flows with the same id seem to be scheduled on the same node which
used to cause panics. How that can happen is currently unclear; cosmic
rays? This patch turns the panic into an RPC or query error (depending
on sync/async flow), adds some more sanity checks and adds a
representation of the flow to the error and the error will also invite
users to report on #12876.

Touches #12876.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17876)
<!-- Reviewable:end -->

RaduBerinde commented on pkg/sql/distsqlrun/flow.go:
```diff
@@ -347,7 +352,12 @@ func (f *Flow) setup(ctx context.Context, spec *FlowSpec) error {
+//
+// Generally if errors are encountered during the setup part, they're returned.
+// But if the flow is a synchronous one, then no error is returned; instead the
+// setup error is pushed to the syncFlowConsume. If an error is returned, a
```
syncFlowConsumer

RaduBerinde commented on pkg/sql/distsqlrun/flow.go:
```diff
@@ -358,18 +368,32 @@ func (f *Flow) Start(ctx context.Context, doneFn func()) {
+			// For sync flows, the error goes to the consumer.
+			f.syncFlowConsumer.Push(nil /* row */, ProducerMetadata{Err: err})
+			f.syncFlowConsumer.ProducerDone()
+		} else {
```
[nit] would be cleaner to invert the condition

RaduBerinde reviewed (approved):
LGTM, thanks!

andreimatei commented:
Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/distsqlrun/flow.go, line 358 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17876#-KsG-rj6g6BfJtr6wvMF-r1-358:-KsGFd_DCWM9yTdmM06h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/015c672d0026b74b883d43baae952f7ecaba53db/pkg/sql/distsqlrun/flow.go#L358)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

syncFlowConsumer
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/flow.go, line 382 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17876#-KsG-rj6g6BfJtr6wvMF-r1-382:-KsGFeLm3paopVl77Usy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/015c672d0026b74b883d43baae952f7ecaba53db/pkg/sql/distsqlrun/flow.go#L382)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] would be cleaner to invert the condition
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17876)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/distsqlrun/flow.go, line 382 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17876#-KsG-rj6g6BfJtr6wvMF-r1-382:-KsGP_HRjUZ2aMCm7yeQ:b-ls8a5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/015c672d0026b74b883d43baae952f7ecaba53db/pkg/sql/distsqlrun/flow.go#L382)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

actually I left the condition like it but this actually needed a `return` on the `syncFlowConsumer != nil` case. Which means that the other branch is not unindented - and I think that's good since it's the more common branch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17876)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
distsqlrun: include the gateway node in the FlowSpec

So it can be part of the error message generated on "flow already
registered" errors. Coupled with the flow registry's node id (which is
now also printed), this should give us info about whether the error has
been generated on the gateway or on a remote node.
